src/saveload/ser.rs

```diff
@@ -0,0 +1,169 @@
+use std::fmt::Display;
+use std::marker::PhantomData;
+
+use serde::ser::{self, Serialize, SerializeSeq, Serializer};
+
+use {Entities, FetchMut, Join, ReadStorage, WriteStorage};
+
+use super::{Components, EntityData, Storages};
+use super::marker::{Marker, MarkerAllocator};
+
+/// Serialize components from specified storages via `SerializableComponent::save`
+/// of all marked entities with provided serializer.
+/// When the component gets serialized with `SerializableComponent::save` method
+/// closure passed in `ids` argument mark unmarked `Entity` and they get serialized recursively.
+/// For serializing without such recursion see `serialize` function.
+pub fn serialize_recursive<'a, M, E, T, S>(
+    entities: &Entities<'a>,
+    storages: &<T as Storages<'a>>::ReadStorages,
+    markers: &mut WriteStorage<'a, M>,
+    allocator: &mut FetchMut<'a, M::Allocator>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    M: Marker,
+    E: Display,
+    T: Components<M::Identifier, E>,
+    S: Serializer,
+{
+    let mut serseq = serializer.serialize_seq(None)?;
+    let mut to_serialize = (&**entities, &*markers)
+        .join()
+        .map(|(e, m)| (e, *m))
+        .collect::<Vec<_>>();
+    loop {
+        if to_serialize.is_empty() {
```

Comment:
any reason to not write it as `while !to_serialize.is_empty() {...}`

Reply:
No reason now.